### PR TITLE
tcc_win: fix tests

### PR DESCRIPTION
**Pull request by spaceface777**
This PR aims to fix the compilation of all test files under `tcc` on Windows.   
Hopefully, once they all pass, we can add a `windows-tcc` CI job to ensure it doesn't break again.

### CODE REVIEW

**Comment by SleepyRoy** [comment]
Amazing! With v+tcc we're going to have the most portable modern language compiler on windows.

**Comment by SleepyRoy** [comment]
I've basically understood the problem with math and solution to it.
Reference: https://github.com/raysan5/raylib/issues/435
Back to our case, the following 7 functions suffer from `undefined symbol` error: `cbrt`, `erf`, `erfc`, `exp2`, `tgamma`, `log2`, `lgamma`. But any V code importing math doesn't work.
In short, take `erf` as an example, one solution is changing in tcc's `math.h`
```C
extern double __cdecl erf (double);
```
to your own solution
```C
static double erf (double x) {return IMPLEMENTATION_OF_ERF(x);}
```
However the `IMPLEMENTATION_OF_ERF` here cannot simply be calling `erf` (compiling is OK but running would give `unhandled exception`). Seems the only way is copying of pasting the implementation of it and I'm still working on this...

**Comment by SleepyRoy** [comment]
OK. Here's so far my best solution to math.

A.
Later I'll open a PR on tccwin_bin repo to change the seven problematic definitions such as  
```C
extern double __cdecl erf (double);
```
to
```C
static double erf (double x) {return erf(x);}
```
By this there're won't be undefined symbol errors.

B.
I suggest for `cbrt`, `exp2`, `log2` the three elementary functions, you change their V math module implementations under win+tcc to `pow(x, 1.0/3)`, `pow(2, x)`, `log10(x)/log10(2)` respectively as workarounds. 
For the rest four non-elementary functions `erf`, `erfc`, `lgamma`, `tgamma` probably you have to print 'sorry not implemented for win+tcc yet' because they in contrast don't have workarounds. Also don't forget to add some `$if` around these functions in math_test, otherwise tests would fail once win+tcc CI is there. 

By doing this, win+tcc users would be able to use all math functions except for the four non-elementary ones.

**Comment by spaceface777** [comment]
Found some more info on the issue:
https://lists.gnu.org/archive/html/tinycc-devel/2017-03/msg00022.html
https://lists.gnu.org/archive/html/tinycc-devel/2017-03/msg00023.html

TL;DR:
- tcc for Windows does not bundle `libm`, which is where these are defined

We may be able to bundle / link against our own dll, will keep looking.

**Comment by SleepyRoy** [comment]
About `string_interpolation_test`, interpolation for large ints is incorrect and here're some clues.
```V
fn main() {
    a := 1000
    b := i64(4294967295)
    c := i64(4294967296)
    d := i64(98765432109876543210)
    println('$a')  // correctly print 1000
    println('$b')  // correctly print 4294967296
    println('$c')  // wrontly print 0 ( c % 4294967296(
    println('$d')  // wrongly print 3844505322 (d % 4294967296)
}
```

**Comment by spytheman** [comment]
The printing depends on  PRId64 , defined in vlib/v/gen/cheaders.v , to be:
`#define PRId64 "lld"`
It probably needs a redefinition for windows's tcc's libc printf.

**Comment by SleepyRoy** [comment]
I just managed to fix http_test and cookie_test. 
Steps:
1. (Add headers) Download full winapi from tcc official - http://download.savannah.gnu.org/releases/tinycc/winapi-full-for-0.9.27.zip. Then unzip it and copy and paste all headers in winapi folder to tccwin's include folder.
2.  When testing, add user32, i.e. use `v -cc tcc: -luser32 test XXX.v`.

**Comment by spaceface777** [comment]
> I just managed to fix http_test and cookie_test. Steps:
> 
> 1. (Add headers) Download full winapi from tcc official - http://download.savannah.gnu.org/releases/tinycc/winapi-full-for-0.9.27.zip. Then unzip it and copy and paste all headers in winapi folder to tccwin's include folder.
> 2. When testing, add user32 flag, i.e. use `v -cc tcc: -luser32 test XXX.v`.

Nice job! Didn't know there was an existing winapi for tcc, it could have saved me some time :)

**Comment by SleepyRoy** [comment]
image

**Comment by spaceface777** [comment]
@SleepyRoy Can't get the tests to pass myself, getting a segfault.
Also, can you post a PR in `vlang/tccbin_win` with the new headers?

**Comment by SleepyRoy** [comment]
@spaceface777 
PR posted. And I just confirmed it worked. Here I give full steps.
1. Download master V and VC.
2. Download tcc from my PR https://github.com/SleepyRoy/tccbin_win/tree/Roy-add-full-winapi.
3. Add PATH.
4. Build. `tcc -std=c99 -municode -lws2_32 -lshell32 -ladvapi32 -bt10 -w -o v.exe v_win.c`
5. Bootstrap. `v -cc tcc self`
6. Test. `v -cc tcc: -luser32 test XXX.v`

**Comment by SleepyRoy** [comment]
image

**Comment by spaceface777** [comment]
@SleepyRoy Do you have another C compiler in your PATH? 
```
v -cc tcc: (...)
         ^
```

that is changing the compiler name, so it is not `tcc`. V is probably falling back to another C compiler to build the test, that's probably why its passing.

Anyways, when fixing the v command, I get this error:
```
> v -cc tcc -cflags "-luser32" test vlib\net\http\http_test.v
(...)
tcc: error: undefined symbol 'WSAConnectByName'
(...)
```
Think I have a fix for it, though, will update soon...

**Comment by spaceface777** [comment]
Update: 

1. The `net/http` tests pass now 🙂
2. The `math` tests "pass" too (after skipping the unimplemented parts in `math_test.v`,   
and updating the signatures of the missing functions in `include/math.h` to 
`static double erf (double x) { return 0; };`). These still need a real implementation, though.

**Comment by SleepyRoy** [comment]
@spaceface777 Nice work! Yeah, I don't have a better idea for math.
And I just posted another PR adding only necessary winapi to pass http. This helps tcc+v preserve its benefits of 'lightweight'. Full winapi is really large for tcc such a mini compiler.
https://github.com/vlang/tccbin_win/pull/5

**Comment by SleepyRoy** [comment]
@spaceface777
So I just tried V with this PR + tccwin with my two PRs (https://github.com/vlang/tccbin_win/pull/3, https://github.com/vlang/tccbin_win/pull/6). All test-fixed pass.

**Comment by spaceface777** [comment]
> @spaceface777
> So I just tried V with this PR + tccwin with my two PRs ([vlang/tccbin_win#3](https://github.com/vlang/tccbin_win/pull/3), [vlang/tccbin_win#6](https://github.com/vlang/tccbin_win/pull/6)). All test-fixed pass.

@SleepyRoy The math tests are only passing because the broken parts are being skipped.  
As I said in vlang/tccbin_win#3, I'm working on building `fdlibm`, so when I'm done the skips can be removed and all tests *will* pass with `tcc`.

After that's done, I think we should add a `tcc-win` CI job before this is merged, to prevent tcc from breaking again on Windows.

**Comment by medvednikov** [comment]
for sure, we'll add a tcc ci

**Comment by spaceface777** [comment]
image

**Comment by spaceface777** [comment]
The new CI (with the `windows-tcc` job) is green now, this can be merged. 🙂

**Comment by Delta456** [comment]
@spaceface777 amazing work man! I will be able to use V on windows as well!

**Comment by medvednikov** [comment]
Great job!
